cheroot/workers/monitor.py

```diff
@@ -0,0 +1,131 @@
+"""Monitor for threadpool."""
+__metaclass__ = type
+
+import threading
+import time
+
+from dynpool import DynamicPoolResizer
+
+
+class BackgroundTask(threading.Thread):
+    """A subclass of threading.Thread whose run() method repeats.
+
+    Use this class for most repeating tasks. It uses time.sleep() to wait
+    for each interval, which isn't very responsive; that is, even if you call
+    self.cancel(), you'll have to wait until the sleep() call finishes before
+    the thread stops. To compensate, it defaults to being daemonic, which means
+    it won't delay stopping the whole process.
+
+    :param int interval: time interval between invocation of input function
+    :param function: callable that needs to be invoked
+    :param args: list of arguments to pass to the function
+    :param kwargs: dictionary of keyword args to pass to the function
+    """
+
+    def __init__(self, interval, function, args=None, kwargs=None):
+        """Initialize a BackgroundTask."""
+        super(BackgroundTask, self).__init__()
+        self.interval = interval
+        self.function = function
+        self.args = args if args is not None else []
+        self.kwargs = kwargs if kwargs is not None else {}
+        self.running = False
+        self.daemon = True
+
+    def cancel(self):
+        """Cancel the task."""
+        self.running = False
+
+    def run(self):
+        """Invoke the callable with the given interval."""
+        self.running = True
+        while self.running:
+            time.sleep(self.interval)
+            if not self.running:
+                return
+            # Any error in the calling function will be raised immediatly
+            self.function(*self.args, **self.kwargs)
+
+
+class Monitor:
+    """Monitor a BackgroundTask."""
+
+    callback = None
+    """The function to call at intervals."""
+
+    frequency = 60
+    """The time in seconds between callback runs."""
+
+    thread = None
+    """A :class:`BackgroundTask` thread."""
+    def __init__(self, callback, frequency=60, name=None):
+        """Initialize Monitor."""
+        self.callback = callback
+        self.frequency = frequency
+        self.thread = None
+        self.name = name or self.__class__.__name__
+
+    def start(self):
+        """Start our callback in its own background thread."""
+        if self.frequency <= 0:
+            return
+        if self.thread is not None:
+            return
+        self.thread = BackgroundTask(self.frequency, self.callback)
+        self.thread.setName(self.name)
```

Comment:
I think it's reasonable to move it inside `BackgroundTask.__init__()` because the initializer should produce a ready-to-use object rather than requiring extra init steps.

Reply:
Done